waku/v2/protocol/waku_noise/noise_utils.nim

```diff
@@ -406,4 +406,9 @@ proc deserializePayloadV2*(payload: seq[byte]): Result[PayloadV2, cstring]
   payload2.transportMessage = payload[i..i+transportMessageLen-1]
   i += transportMessageLen
 
-  return ok(payload2)
\ No newline at end of file
+  return ok(payload2)
+
+# Simple utility that checks if the given variable is "default", 
+# Therefore, it has not been initialized
+proc isDefault*[T](defaultyVar: T): bool = 
```

Comment:
Please, either turn this into a `{.inline.}` _proc_ or turn this into a _template_. So we ensure that this is optimized at compile time. Personally, I prefer the _template_ option.

Reply:
The rationale for using an _inline proc_ or _template_ is because if you use a normal _proc_ it is up to the compiler to optimize it. 

Using a normal proc, if it is not optimized by the nim/C compiler, may mean that you are creating a new stack frame for just doing a comparison every time you do this comparison. If you use a template/inline proc, at compile time the comparison is embedded in the statement code during the code preprocessing.

Reply:
Here a thread from the Nim forum about _inline proc vs template_: https://forum.nim-lang.org/t/3948

Reply:
Resolved in [fb2b652](https://github.com/status-im/nwaku/pull/1104/commits/fb2b6528d0b1c3ea324df31e3e8c19b7498314fb)

Reply:
we generally avoid `inline`: https://status-im.github.io/nim-style-guide/language.inline.html (LTO is better almost always)

Reply:
FWIW @s1fr0 Check the heavy usage of the templates in _nim-stew's results_ package 😁
https://github.com/status-im/nim-stew/blob/master/stew/results.nim

Reply:
lol, sorry to crash the party unannounced .. that said, `proc` without inline is usually fine - LTO will inline it at the linking stage, if it's suitable for inlining (small enough and not recursive) - if I was worried about performance, I'd be more worried about `default(T)` which has to create an all-zero instance on the fly which can be slow if `T` is large - but... it's rare for performance problems to be found in these kind of functions / cases because they don't get called often enough that it matters - `proc` is the safer choice as you correctly point out above, because `template` has some gotchas, and until proven otherwise with strong reasoning or a profiler, safety is more interesting

Reply:
>  Check the heavy usage of the templates

so - templates are useful sometimes indeed - from a functional perspective, you get access to symbols from the instantiation context which in the case of `nim-result` is necessary to deduce the type of the special `result` variable - the other case is when dealing with return types: functions that return large types have historically been inefficient, and some of the templates in nim-result sometimes "return" large types - finally, some of the nim-result code was written pre-LTO so it still sort of lives with that legacy. 

Now, there's plenty of nuance if you go looking for it, _but_ in cases like this it wouldn't make much difference no matter what you do - compilers are pretty clever these days, and profilers are even better - more importantly though, as a matter of style, this should be a `func` to indicate that it is side-effect free - this helps future developers trust that the function won't do weird things.

Reply:
for default, there's a trick I think - one can do `static(default(T))` and then a constant will be generated such that the runtime zero:ing of memory is avoided

Reply:
@arnetheduck thanks for your insight, addressed in [a42e509](https://github.com/status-im/nwaku/pull/1104/commits/a42e509cf242956159133d9d38974afe6b1e72b0)

Reply:
no problem :)

if you want to continue tuning the code, switching most `proc` to `func` will help the compiler help you in the future -  the next step is to not rely on `default` at all, but rather use a type like `results.Opt` which keeps track of "unset" values more clearly (as far as the reader is concerned). 

For many types of values, depending on the serialization format, all-zeroes _might_ be a valid object (a keypair is unlikely to suffer from this, but...) - `Opt` removes the risk of falling into this trap.